src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+        /// </summary>
+        public void WriteUndefined()
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.UndefinedValue);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteStartAttribute(string name)
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+
+            WriteSpacing();
+            WriteQuotedString(name);
+            WriteChar(JsonConstants.KeyValueSeperator);
+
+            if (_prettyPrint)
+                WriteChar(JsonConstants.Space);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void WriteNewline()
+        private void WriteChar(char value)
         {
-            if (PrettyPrint) {
-                _formatter.Append('\n');
+            var output = _output;
+            if (UseFastUtf8)
+            {
+                if (output.Buffer.Length < 1)
+                    output.Enlarge();
+
+                output.Buffer.DangerousGetPinnableReference() = (byte)value;
+                output.Advance(1);
             }
+            else if (UseFastUtf16)
+            {
+                if (output.Buffer.Length < 2)
+                    output.Enlarge(2);
+
+                unsafe
+                {
+                    Unsafe.AsRef<char>(Unsafe.AsPointer<byte>(ref output.Buffer.DangerousGetPinnableReference())) = value;
+                }
+
+                output.Advance(2);
+            }
+            else
+            {
+                unsafe
+                {
+                    // Slow path, if we are dealing with non-invariant.
+                    ReadOnlySpan<char> chSpan = new ReadOnlySpan<char>(&value, 1);
+                    Write(chSpan);
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteQuotedString(string value)
+        {
+            WriteChar(JsonConstants.Quote);
+            // TODO: We need to handle escaping.
+            Write(value.AsSpan());
+            WriteChar(JsonConstants.Quote);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNumber(long value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.NumberFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteNumber(ulong value)
+        {
+            int written;
+            while (!value.TryFormat(_output.Buffer, out written, JsonConstants.NumberFormat, _output.Encoder))
+                _output.Enlarge();
+
+            _output.Advance(written);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteDateTime(DateTime value)
```

Comment:
None of the types I added overloads for actual allocate. They use the primitive formatter for all of this. So, though I said "writes DateTime as a string", it doesn't actually convert it to a string then write it.

The write methods are all there to stay consistent with JSON.NET. Enough people write dates and guids to JSON that these make sense. They have the benefit of being non-allocating. If they weren't there, most people would likely do a string convert and write the string. In the "efficient" case, they would still need a temporary byte buffer / Span to use the primitive formatter first.